Mesh: Add lattice free-form deformation (#1814)
NVIDIA/physicsnemo: docs/api/mesh/transformations.rst

   :alt: Original sphere and single-control and multiple-control sphere morphing
   :width: 100%

Lattice Free-Form Deformation
^^^^^^^^^^^^^^^^^^^^^^^^^^^^^

:meth:`~physicsnemo.mesh.mesh.Mesh.ffd` defines a regular array of control
displacements over an axis-aligned evaluation box and deforms every point
inside the box by tensor-product basis interpolation.
Compared with sparse morphing, the design parameters form a structured grid of

> fixed size, which suits parametric shape optimization: a lattice of zeros is

Reviewer: ```suggestion
fixed size, which suits parametric shape optimization. A lattice of zeros is
```

exactly the identity, and the same lattice deforms any geometry embedded in
the box.

> When ``origin`` and ``extent`` are omitted, the box spans the mesh bounds. Each

Reviewer: i am trying to leave a comment on line 171

coordinate axis must have positive range; planar or linear geometry embedded
in a higher-dimensional space therefore needs an explicit positive extent.

> ``basis="bernstein"`` is classic global-support free-form deformation for

Reviewer: ```suggestion
``basis="bernstein"`` provides classic global-support free-form deformation for
```

coarse lattices. ``basis="bspline"`` gives local four-node-per-axis support

> and scales to fine lattices for local sculpting; its first and last coefficient

Reviewer: ```suggestion
and scales to fine lattices for local sculpting. Its first and last coefficient
```

planes lie one knot spacing outside the evaluation box. ``basis="linear"``,
``"smoothstep"``, and ``"smootherstep"`` instead use the two neighboring
nodes per axis and reproduce every control displacement at its lattice node.
Their cell transitions are respectively C0, C1, and C2.

.. code:: python

    # A 4x4x4 Bernstein lattice spanning the mesh bounds; zero displacements
    # start at the identity.
    control_displacements = torch.zeros(4, 4, 4, 3, requires_grad=True)
    deformed = mesh.ffd(control_displacements)

    # Autograd continues through the returned point coordinates.
    objective = deformed.points.square().mean()
    objective.backward()

Points outside the lattice box are unchanged. The deformation is generally not
continuous across the box boundary. A sufficient condition for a fixed
exterior is to zero the outermost coefficient plane on every Bernstein or

> node-interpolating face, or the first and last three coefficient planes on

Reviewer: ```suggestion
node-interpolating face. For cubic B-splines, zero the first and last three coefficient planes on
```

> every axis for cubic B-splines. ``origin`` and ``extent`` are

Reviewer: ```suggestion
every axis.. ``origin`` and ``extent`` are
```

> non-differentiable lattice configuration; optimize ``control_displacements``

Reviewer: ```suggestion
non-differentiable lattice parameters. Optimize ``control_displacements``
```

        morphed_domain.boundaries["wall"].points,
    )

:meth:`~physicsnemo.mesh.domain_mesh.DomainMesh.ffd` follows the same pattern

> for lattice free-form deformation: one lattice field is evaluated over the

Reviewer: ```suggestion
for lattice free-form deformation: 
```

> combined interior and boundary points, and the default box spans the combined

Reviewer: ```suggestion

- One lattice field is evaluated over the combined interior and boundary points.
- The default box spans the combined component bounds.
```

> component bounds. The combined bounds must have positive range on every

Reviewer: ```suggestion

```

> coordinate axis unless an explicit extent is supplied.

Reviewer: ```suggestion
The combined bounds must have positive range on every coordinate axis unless an
explicit extent is supplied.
```

Every deformation preserves connectivity and attached point, cell, global, and

> domain data. Attached vector and tensor fields are treated as Lagrangian data

Reviewer: ```suggestion
domain data. The mesh treats attached vector and tensor fields as Lagrangian
```

> and are not pushed forward. Geometry-dependent caches are discarded and

Reviewer: ```suggestion
data and does not push them forward. Geometry-dependent caches are discarded and
```

> recomputed lazily; topology caches are retained.

Reviewer: ```suggestion
recomputed lazily. Topology caches are retained.
```
